consensus/raft/raft.go

```diff
@@ -5,173 +5,239 @@ import (
 	if err != nil {
-		logger.Error("creating libp2p-raft transport: ", err)
 		return nil, err
 	}
+	srvCfg := makeServerConf(peers)
 
-	pstore := &libp2praft.Peerstore{}
-	peersStr := make([]string, len(peers), len(peers))
-	for i, p := range peers {
-		peersStr[i] = peer.IDB58Encode(p)
-	}
-	pstore.SetPeers(peersStr)
-
-	logger.Debug("creating file snapshot store")
-	dataFolder := cfg.DataFolder
-	if dataFolder == "" {
-		dataFolder = filepath.Join(cfg.BaseDir, DefaultDataSubFolder)
+	logger.Debug("creating libp2p Raft transport")
+	transport, err := p2praft.NewLibp2pTransport(host, cfg.NetworkTimeout)
+	if err != nil {
+		return nil, err
 	}
 
-	err = os.MkdirAll(dataFolder, 0700)
+	logger.Debug("creating raft snapshot store")
+	snapshots, err := hraft.NewFileSnapshotStoreWithLogger(
+		dataFolder, RaftMaxSnapshots, raftStdLogger)
 	if err != nil {
-		logger.Errorf("creating cosensus data folder (%s): %s",
-			dataFolder, err)
 		return nil, err
 	}
-	snapshots, err := hashiraft.NewFileSnapshotStoreWithLogger(dataFolder, RaftMaxSnapshots, raftStdLogger)
+
+	logger.Debug("creating BoltDB log store")
+	logStore, err := raftboltdb.NewBoltStore(
+		filepath.Join(dataFolder, "raft.db"))
 	if err != nil {
-		logger.Error("creating file snapshot store: ", err)
 		return nil, err
 	}
 
-	logger.Debug("creating BoltDB log store")
-	logStore, err := raftboltdb.NewBoltStore(filepath.Join(dataFolder, "raft.db"))
+	logger.Debug("checking for existing raft states")
+	hasState, err := raft.HasExistingState(logStore, logStore, snapshots)
 	if err != nil {
-		logger.Error("creating bolt store: ", err)
 		return nil, err
 	}
+	if !hasState {
+		logger.Info("bootstrapping raft cluster")
+		err := raft.BootstrapCluster(cfg.RaftConfig,
+			logStore, logStore, snapshots, transport, srvCfg)
+		if err != nil {
+			logger.Error("bootstrapping cluster: ", err)
+			return nil, err
+		}
+	} else {
+		logger.Info("raft cluster is already bootstrapped")
+	}
 
 	logger.Debug("creating Raft")
-	r, err := hashiraft.NewRaft(cfg.HashiraftCfg, fsm, logStore, logStore, snapshots, pstore, transport)
+	r, err := hraft.NewRaft(cfg.RaftConfig,
+		fsm, logStore, logStore, snapshots, transport)
 	if err != nil {
 		logger.Error("initializing raft: ", err)
 		return nil, err
 	}
 
-	raft := &Raft{
+	raftW := &raftWrapper{
 		raft:          r,
+		srvConfig:     srvCfg,
 		transport:     transport,
 		snapshotStore: snapshots,
 		logStore:      logStore,
 		stableStore:   logStore,
-		peerstore:     pstore,
 		boltdb:        logStore,
-		dataFolder:    dataFolder,
 	}
 
-	return raft, nil
+	// Handle existing, different configuration
+	if hasState {
+		logger.Error("has state")
+		cf := r.GetConfiguration()
+		if err := cf.Error(); err != nil {
+			return nil, err
+		}
+		currentCfg := cf.Configuration()
+		added, removed := diffConfigurations(srvCfg, currentCfg)
+		if len(added)+len(removed) > 0 {
+			raftW.Shutdown()
+			logger.Error("Raft peers do not match cluster peers")
```

Comment:
In the future, or now...

you mean setting `peers` to the right value and move on? Or nuking Raft data directly?

If peers have changed and Raft doesn't expect that, this is a problem and very likely we would be merging diverging states. Nuking the data is risky as it may be valuable.

I have in mind that we need to have things `ipfs-cluster-service clean` and/or `ipfs-cluster-service reset`. In short:

* A way to remove all Raft data and leave cluster in a clean state
* A way to start a cluster peer which had peers before in single peer-mode, without losing the data
* A way to backup and re-import data from a dirty peer  to a clean one.

But all these are self-contained and can be approached on different PRs.